### PR TITLE
Use optimized_transducer to compute transducer loss.

### PR DESCRIPTION
**Pull request by csukuangfj**
Use https://github.com/csukuangfj/optimized_transducer to compute the transducer loss and remove the dependency on torchaudio 0.10.0.

https://github.com/csukuangfj/optimized_transducer is based on torchaudio and implements the optimization techniques
proposed in "Improving RNN Transducer Modeling for End-to-End Speech Recognition" (https://arxiv.org/abs/1909.12415)

---

[EDITED]: Need to test
- [ ] training speed
- [ ] memory consumption
- [ ] WER

### CODE REVIEW

**Comment by csukuangfj** [comment]
Here are some training logs comparing this PR and the master.

## This PR with [optimized_transducer](https://github.com/csukuangfj/optimized_transducer)
```
2021-12-30 00:41:27,413 INFO [train.py:648] epoch 0, learning rate 1.111738239902946e-06
2021-12-30 00:41:29,155 INFO [train.py:492] Epoch 0, batch 0, loss[loss=6.719, over 4855 frames.], tot_loss[loss=6.719, over 4855 frames.], batch size: 14
2021-12-30 00:42:05,497 INFO [train.py:492] Epoch 0, batch 50, loss[loss=2.714, over 4974 frames.], tot_loss[loss=4.615, over 218604.36685246887 frames.], batch size: 16
2021-12-30 00:42:41,468 INFO [train.py:492] Epoch 0, batch 100, loss[loss=1.264, over 4678 frames.], tot_loss[loss=3.027, over 384201.4252568698 frames.], batch size: 13
2021-12-30 00:43:17,561 INFO [train.py:492] Epoch 0, batch 150, loss[loss=1.37, over 4799 frames.], tot_loss[loss=2.285, over 513208.9591736901 frames.], batch size: 15
2021-12-30 00:43:53,482 INFO [train.py:492] Epoch 0, batch 200, loss[loss=1.147, over 4996 frames.], tot_loss[loss=1.91, over 614683.7748667101 frames.], batch size: 13
2021-12-30 00:44:30,064 INFO [train.py:492] Epoch 0, batch 250, loss[loss=1.046, over 4716 frames.], tot_loss[loss=1.682, over 693857.4948371359 frames.], batch size: 11
2021-12-30 00:45:06,011 INFO [train.py:492] Epoch 0, batch 300, loss[loss=1.199, over 4822 frames.], tot_loss[loss=1.537, over 755366.5885046537 frames.], batch size: 16
2021-12-30 00:45:42,284 INFO [train.py:492] Epoch 0, batch 350, loss[loss=1.243, over 4858 frames.], tot_loss[loss=1.443, over 801682.4665612186 frames.], batch size: 25
2021-12-30 00:46:18,568 INFO [train.py:492] Epoch 0, batch 400, loss[loss=1.265, over 4935 frames.], tot_loss[loss=1.376, over 836680.4645405153 frames.], batch size: 52
2021-12-30 00:46:54,609 INFO [train.py:492] Epoch 0, batch 450, loss[loss=1.133, over 4925 frames.], tot_loss[loss=1.326, over 863943.9577014449 frames.], batch size: 13
2021-12-30 00:47:30,546 INFO [train.py:492] Epoch 0, batch 500, loss[loss=1.086, over 4666 frames.], tot_loss[loss=1.292, over 886262.7250945688 frames.], batch size: 12
2021-12-30 00:48:06,796 INFO [train.py:492] Epoch 0, batch 550, loss[loss=1.233, over 4909 frames.], tot_loss[loss=1.265, over 904647.3838964644 frames.], batch size: 17
2021-12-30 00:48:43,591 INFO [train.py:492] Epoch 0, batch 600, loss[loss=1.131, over 4989 frames.], tot_loss[loss=1.24, over 919067.3875420308 frames.], batch size: 13
2021-12-30 00:49:19,788 INFO [train.py:492] Epoch 0, batch 650, loss[loss=1.185, over 4859 frames.], tot_loss[loss=1.225, over 930203.6480491284 frames.], batch size: 13
2021-12-30 00:49:56,062 INFO [train.py:492] Epoch 0, batch 700, loss[loss=1.099, over 4619 frames.], tot_loss[loss=1.206, over 937584.9792811682 frames.], batch size: 11
2021-12-30 00:50:32,230 INFO [train.py:492] Epoch 0, batch 750, loss[loss=1.043, over 4971 frames.], tot_loss[loss=1.198, over 943856.7463587269 frames.], batch size: 11
2021-12-30 00:51:08,590 INFO [train.py:492] Epoch 0, batch 800, loss[loss=1.131, over 4975 frames.], tot_loss[loss=1.189, over 949027.5236152575 frames.], batch size: 16
2021-12-30 00:51:08,591 INFO [train.py:509] Computing validation loss
2021-12-30 00:51:46,426 INFO [train.py:518] Epoch 0, validation: loss=1.186, over 944034 frames.
2021-12-30 00:52:22,813 INFO [train.py:492] Epoch 0, batch 850, loss[loss=1.117, over 4795 frames.], tot_loss[loss=1.186, over 953485.5502514712 frames.], batch size: 13
2021-12-30 00:52:58,973 INFO [train.py:492] Epoch 0, batch 900, loss[loss=1.28, over 4780 frames.], tot_loss[loss=1.178, over 955710.1403901987 frames.], batch size: 15
```

## master with torchaudio
```
2021-12-30 00:41:29,152 INFO [train.py:648] epoch 0, learning rate 1.111738239902946e-06
2021-12-30 00:41:30,611 INFO [train.py:492] Epoch 0, batch 0, loss[loss=6.719, over 4855 frames.], tot_loss[loss=6.719, over 4855 frames.], batch size: 14
2021-12-30 00:42:08,694 INFO [train.py:492] Epoch 0, batch 50, loss[loss=2.714, over 4974 frames.], tot_loss[loss=4.615, over 218604.36685246887 frames.], batch size: 16
2021-12-30 00:42:46,392 INFO [train.py:492] Epoch 0, batch 100, loss[loss=1.264, over 4678 frames.], tot_loss[loss=3.027, over 384201.4252568698 frames.], batch size: 13
2021-12-30 00:43:24,252 INFO [train.py:492] Epoch 0, batch 150, loss[loss=1.37, over 4799 frames.], tot_loss[loss=2.285, over 513208.9591736901 frames.], batch size: 15
2021-12-30 00:44:02,114 INFO [train.py:492] Epoch 0, batch 200, loss[loss=1.147, over 4996 frames.], tot_loss[loss=1.91, over 614683.7748667101 frames.], batch size: 13
2021-12-30 00:44:40,570 INFO [train.py:492] Epoch 0, batch 250, loss[loss=1.046, over 4716 frames.], tot_loss[loss=1.682, over 693857.4948371359 frames.], batch size: 11
2021-12-30 00:45:18,271 INFO [train.py:492] Epoch 0, batch 300, loss[loss=1.199, over 4822 frames.], tot_loss[loss=1.537, over 755366.5885046537 frames.], batch size: 16
2021-12-30 00:45:56,014 INFO [train.py:492] Epoch 0, batch 350, loss[loss=1.243, over 4858 frames.], tot_loss[loss=1.443, over 801682.4665612186 frames.], batch size: 25
2021-12-30 00:46:33,901 INFO [train.py:492] Epoch 0, batch 400, loss[loss=1.265, over 4935 frames.], tot_loss[loss=1.376, over 836680.4645405153 frames.], batch size: 52
2021-12-30 00:47:11,847 INFO [train.py:492] Epoch 0, batch 450, loss[loss=1.133, over 4925 frames.], tot_loss[loss=1.326, over 863943.9577014449 frames.], batch size: 13
2021-12-30 00:47:49,525 INFO [train.py:492] Epoch 0, batch 500, loss[loss=1.086, over 4666 frames.], tot_loss[loss=1.292, over 886262.7250945688 frames.], batch size: 12
2021-12-30 00:48:27,316 INFO [train.py:492] Epoch 0, batch 550, loss[loss=1.233, over 4909 frames.], tot_loss[loss=1.265, over 904647.3838964644 frames.], batch size: 17
2021-12-30 00:49:05,743 INFO [train.py:492] Epoch 0, batch 600, loss[loss=1.131, over 4989 frames.], tot_loss[loss=1.24, over 919067.3875420308 frames.], batch size: 13
2021-12-30 00:49:43,712 INFO [train.py:492] Epoch 0, batch 650, loss[loss=1.185, over 4859 frames.], tot_loss[loss=1.303, over 930203.6480491284 frames.], batch size: 13
2021-12-30 00:50:21,902 INFO [train.py:492] Epoch 0, batch 700, loss[loss=1.099, over 4619 frames.], tot_loss[loss=1.266, over 937584.9792811682 frames.], batch size: 11
2021-12-30 00:50:59,670 INFO [train.py:492] Epoch 0, batch 750, loss[loss=1.043, over 4971 frames.], tot_loss[loss=1.245, over 943856.7463587269 frames.], batch size: 11
2021-12-30 00:51:37,519 INFO [train.py:492] Epoch 0, batch 800, loss[loss=1.132, over 4975 frames.], tot_loss[loss=1.225, over 949027.5236152575 frames.], batch size: 16
2021-12-30 00:51:37,520 INFO [train.py:509] Computing validation loss
2021-12-30 00:52:29,845 INFO [train.py:518] Epoch 0, validation: loss=1.187, over 944034 frames.
2021-12-30 00:53:07,777 INFO [train.py:492] Epoch 0, batch 850, loss[loss=1.117, over 4795 frames.], tot_loss[loss=1.214, over 953485.5502514712 frames.], batch size: 13
2021-12-30 00:53:45,523 INFO [train.py:492] Epoch 0, batch 900, loss[loss=1.282, over 4780 frames.], tot_loss[loss=1.199, over 955710.1403901987 frames.], batch size: 15
```

You can see that [optimized_transducer](https://github.com/csukuangfj/optimized_transducer) produces
the same loss as torchaudio and it is also a bit faster.

---

The training command is
```
./transducer_stateless/train.py \
  --world-size 1 \
  --num-epochs 40 \
  --start-epoch 0 \
  --exp-dir transducer_stateless/exp-100 \
  --full-libri 0 \
  --max-duration 200 \
  --lr-factor 3
```

Would be great if someone else can reproduce the above result. (Note: It requires the latest optimized_transducer, i.e., v1.1,
which can be installed with `pip install optimized_transducer`)

**Comment by TeaPoly** [comment]
>  memory consumption

Thanks for @csukuangfj. `optimized_transducer` is great job. Your working has greatly save the memory usage, but because `optimized_transducer` currently only works on `FP32`, it still occupies more memory than `torchaudio` in `FP16`.

  | FP32 | FP16
-- | -- | --
`torchaudio.functional.rnnt_loss` | 8,559 | 4,593
`optimized_transducer.transducer_loss` | 6,663 | 6,191

**Comment by csukuangfj** [comment]
@TeaPoly 

Thanks for your feedback.

I will add FP16 support to it later.

**Comment by danpovey** [comment]
Would we be doing the forward pass in fp16?  I would expect that the numerical roundoff might be quite problematic for the RNN-T loss.

**Comment by csukuangfj** [comment]
Here are the results for this PR.

|                     | this PR            | the master         | comment                                   |
|---------------------|--------------------|--------------------|-------------------------------------------|
| world size          | 4                  | 4                  | Both use V100 with 32-GB RAM              |
| max duration        | 250                | 200                | It causes OOM with 250 for the master     |
| time per epoch      | 2 hours 50 minutes | 2 hours 47 minutes |                                           |
| tensorboard log     | [click here](https://tensorboard.dev/experiment/FNcJnYPETmS5T56ORWkRPQ/#scalars&_smoothingWeight=0)         | [click here](https://tensorboard.dev/experiment/Mjx7MeTgR3Oyr1yBCwjozw/#scalars&_smoothingWeight=0)         |                                           |
| WER-epoch-21-avg-11 | 2.94/7.53          | 3.01/7.59          | test_clean/test_other, with greedy search |

torchaudio uses warp level parallel reduction and is a little bit faster. Their WERs are comparable.

Will train it for more epochs, e.g., 30 epochs, when there are available GPUs.

**Comment by danpovey** [comment]
Great work!!

**Comment by csukuangfj** [comment]
Here are the WERs I get so far:

|               | test-clean | test-other |                           comment |
|---------------|------------|------------|----------------------------------:|
| greedy search | 2.81       | 7.07       | --epoch 34 --avg 16               |
| beam search   | 2.78       | 7.02       | --epoch 34 --avg 16 --beam-size 4 |

It is a little better than the current master.

Will upload a pre-trained model and merge this PR.

**Comment by danpovey** [comment]
Great!!

**Comment by csukuangfj** [comment]
Here are the latest results for this PR:

|                           | test-clean | test-other | comment                                  |
|---------------------------|------------|------------|------------------------------------------|
| greedy search             | 2.77       | 7.07       | --epoch 36, --avg 13, --max-duration 100 |
| beam search (beam size 4) | 2.76       | 6.97       |                                          |

A pre-trained model is uploaded to https://huggingface.co/csukuangfj/icefall-asr-librispeech-transducer-stateless-bpe-500-2022-01-10

Also, the following colab notebook
https://colab.research.google.com/drive/1Rc4Is-3Yp9LbcEz_Iy8hfyenyHsyjvqE?usp=sharing
shows how to download and use the above pre-trained model to transcribe sound files.
